contributors/design-proposals/network/pod-resolv-conf.md

```diff
@@ -0,0 +1,167 @@
+# Custom /etc/resolv.conf
+* Status: pending
+* Version: alpha
+* Implementation owner: Bowei Du <[bowei@google.com](mailto:bowei@google.com)>, Zihong Zheng <[zihongz@google.com](mailto:zihongz@google.com)>
+
+# Overview
+The `/etc/resolv.conf` in a pod is managed by Kubelet and its contents are generated based on `pod.dnsPolicy`. For `dnsPolicy: Default`, `resolv.conf` is copied from the node where the pod is running. If the `dnsPolicy` is `ClusterFirst`, the contents of the resolv.conf is the hosts `resolv.conf` augmented with the following options:
+
+*   Search paths to add aliases for domain names in the same namespace and cluster suffix.
+*   `options ndots` to 5 to ensure the search paths are searched for all potential matches.
+
+The configuration of both search paths and `ndots` results in query amplification of five to ten times for non-cluster internal names. This is due to the fact that each of the search path expansions must be tried before the actual result is found. This order of magnitude increase of query rate imposes a large load on the kube-dns service. At the same time, there are user applications do not need the convenience of the name aliases and do not wish to pay this performance cost.
+
+
+## Existing workarounds
+
+The current work around for this problem is to specify an FQDN for name resolution. Any domain name that ends with a period (e.g. `foo.bar.com.`) will not be search path expanded. However, use of FQDNs is not well-known practice and imposes application-level changes. Cluster operators may not have the luxury of enforcing such a change to applications that run on their infrastructure.
+
+It is also possible for the user to insert a short shell script snippet that rewrites `resolv.conf` on container start-up. This has the same problems as the previous approach and is also awkward for the user. This also forces the container to have additional executable code such as a shell or scripting engine which increases the applications security surface area.
+
+
+# Proposal sketch
+
+Add a new `pod.dnsPolicy: Custom` that allows for user customization of `resolv.conf`.
+
+
+## Pod API example
+
+In the example below, the user wishes to add the pod namespace and a custom expansion to the search path, as they do not use the other name aliases:
+
+```yaml
+# Pod spec
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: ns1
+  name: example
+spec:
+  containers:
+  - name: example
+    image: example
+  dnsPolicy: Custom
+  dnsParams:
+    searchPaths:
+    - $NAMESPACE.svc.$CLUSTER
+    - my.dns.search.suffix
+    - $HOST
+    options:
+    - ndots: 2
```

Comment:
unclear what the type behind this is?  Is it a map of string-> string?  I think you meant it to just be []string, but I bet this doesn't unmarshal properly (since you wrote a []struct{ndots int}).

you either need to say:

```
options:
- "ndots: 2"
```

or we need to actually make a schema (which I don't think we should do) or we need to make it a map (which we generally frown upon) or we need to make it:

```
options:
- name: ndots
  value: "2"
```

which I dislike.  Assuming no other major issues, we can discuss this as an API issue with API guardians.

Reply:
Want to raise that not all options are {key}:{value}, options like rotate and debug are just flags.  I think []string might be the best because the intention here is to capture the options, and no additional processing is assumed.

Reply:
It is []string (see API code), I updated the YAML to make this clear (it's actually a YAML syntax error)